File: gnsstools/tracking.py
from abc import ABC, abstractmethod
from asyncio import constants
import configparser
import numpy as np
from gnsstools.gnss import GNSS

from gnsstools.gnsssignal import GNSSSignal
from gnsstools.rffile import RFFile
from gnsstools.acquisition import Acquisition
from gnsstools.rfsignal import RFSignal

class Tracking:

    def __init__(self, configfile, acquisition:Acquisition):
        config = configparser.ConfigParser()
        config.read(configfile)

        self.msToProcess = config.getint('DEFAULT', 'ms_to_process')
        
        self.pdiCode    = config.getfloat('TRACKING', 'pdi_code')
        self.pdiCarrier = config.getfloat('TRACKING', 'pdi_carrier')

        # Load tracking parameters
        self.dllCorrelatorSpacing = config.getfloat('TRACKING', 'dll_correlator_spacing')
        self.dllDumpingRatio      = config.getfloat('TRACKING', 'dll_dumping_ratio')
        self.dllNoiseBandwidth    = config.getfloat('TRACKING', 'dll_noise_bandwidth')
        self.dllLoopGain          = config.getfloat('TRACKING', 'dll_loop_gain')

        self.pllDumpingRatio      = config.getfloat('TRACKING', 'pll_dumping_ratio')
        self.pllNoiseBandwidth    = config.getfloat('TRACKING', 'pll_noise_bandwidth')
        self.pllLoopGain          = config.getfloat('TRACKING', 'pll_loop_gain')
        
        self.acquisition = acquisition
        self.prn         = self.acquisition.prn
        self.signal      = self.acquisition.signal
        self.init_freq   = self.acquisition.coarseFreq
        self.init_code   = self.acquisition.coarseCode

        self.absoluteSample   = []
        self.codeFrequency    = []
        self.carrierFrequency = []
        self.codeError        = []
        self.codeNCO          = []
        self.carrierError     = []
        self.carrierNCO       = []
        self.codePhase        = []   
        self.iEarly  = []
        self.qEarly  = []
        self.iPrompt = []
        self.qPrompt = []
        self.iLate   = []
        self.qLate   = []
        
        # Initialise
        self.dllTau1, self.dllTau2 = self.getLoopCoefficients(self.dllNoiseBandwidth, \
            self.dllDumpingRatio, self.dllLoopGain)
        self.pllTau1, self.pllTau2 = self.getLoopCoefficients(self.pllNoiseBandwidth, \
            self.pllDumpingRatio, self.pllLoopGain)

        return
    
    def getLoopCoefficients(self, loopNoiseBandwidth, dumpingRatio, loopGain):
        """
        Compute loop coeficients for PLL and DLL loops. From code of [Borre, 2007].

        Parameters
        ----------
        loopNoiseBandwidth : float
            Loop Noise Bandwith parameter
        dumpingRatio : float
            Dumping Ratio parameter, a.k.a. zeta
        loopGain
            Loop Gain parameter
        
        Returns
        -------
        tau1, tau2 : float, float
            Loop filter coeficients
        """

        Wn = loopNoiseBandwidth * 8.0 * dumpingRatio / \
            (4.0 * dumpingRatio**2 +1)
        
        tau1 = loopGain / Wn**2
        tau2 = 2.0 * dumpingRatio / Wn

        return tau1, tau2

    def track(self, signal_file:RFFile):
        """
        Track signal based on acquisition results. Largely inspired by [Borre, 2017]
        and the Python implementation from perrysou (Github).
        """

        # Generate CA
        caCode = self.signal.getCode(self.prn) # Could be stored in acquisition

        # Extend a bit the code with last and first value
        caCode = np.r_[caCode[-1], caCode, caCode[0]]

        # Initialize
        carrierFrequency = self.init_freq
        codeFrequency    = self.signal.code_freq
        remCarrierPhase  = 0.0 # Keep the remaining part of the carrier phase for next iteration
        remCodePhase     = 0.0 # Keep the remaining part of the code phase for next iteration
        codeNCO          = 0.0
        carrierNCO       = 0.0
        oldCodeError     = 0.0
        oldCarrierError  = 0.0

        # Start tracking
        for code_counter in range(self.msToProcess):
            # -----------------------------------------------------------------
            # Read signal from file
            codePhaseStep = codeFrequency / signal_file.samp_freq
            chunck = int(np.ceil((self.signal.code_bit - remCodePhase) / codePhaseStep))
            
            if code_counter == 0:
                skip = self.init_code + 1
                rawSignal = signal_file.readFileByValues(nb_values=chunck, skip=skip, keep_open=True)
            else:
                rawSignal = signal_file.readFileByValues(nb_values=chunck, keep_open=True)

            # Check if there is enough data in file
            if len(rawSignal) < chunck:
                raise EOFError("EOF encountered earlier than expected in file.")
            
            # -----------------------------------------------------------------
            # Generate the code replica for correlators
            ## Early code 
            idx = np.ceil(np.linspace(remCodePhase - self.dllCorrelatorSpacing, \
                chunck * codePhaseStep + remCodePhase - self.dllCorrelatorSpacing, \
                chunck, endpoint=False)).astype(int)
            
            earlyCode = caCode[idx]

            ## Late code
            idx = np.ceil(np.linspace(remCodePhase + self.dllCorrelatorSpacing, \
                chunck * codePhaseStep + remCodePhase + self.dllCorrelatorSpacing, \
                chunck, endpoint=False)).astype(int)

            lateCode = caCode[idx]

            ## Prompt code
            # We don't apply ceil nor int here because we need the float value
            # of the last index to compute the remaining phase
            idx = np.linspace(remCodePhase, \
                chunck * codePhaseStep + remCodePhase, \
                chunck, endpoint=False)
            
            promptCode = caCode[np.ceil(idx).astype(int)]

            # Update the remain code phase variable
            remCodePhase = idx[chunck-1] + codePhaseStep - self.signal.code_bit

            # -----------------------------------------------------------------
            # Generate carrier replica and mix to remove frequency shift
            # We use (chunck+1) and not (chunck) because we want one more to
            # estimate the remaining of the carrier phase
            time = np.arange(0, chunck+1) / signal_file.samp_freq
            #temp = carrierFrequency * 2.0 * np.pi * time + remCarrierPhase
            temp = -(carrierFrequency * 2.0 * np.pi * time) + remCarrierPhase

            remCarrierPhase = temp[chunck] % (2 * np.pi)
            
            carrierSignal = np.exp(1j * temp[:chunck]) * rawSignal
            iSignal = np.real(carrierSignal)
            qSignal = np.imag(carrierSignal)
            #iSignal = np.sin(temp[:chunck]) * rawSignal # In-phase
            #qSignal = np.cos(temp[:chunck]) * rawSignal # Quadraphase

            # -----------------------------------------------------------------
            # Correlators update
            iEarly  = np.sum(earlyCode  * iSignal)
            qEarly  = np.sum(earlyCode  * qSignal)
            iPrompt = np.sum(promptCode * iSignal)
            qPrompt = np.sum(promptCode * qSignal)
            iLate   = np.sum(lateCode   * iSignal)
            qLate   = np.sum(lateCode   * qSignal)
            
            # -----------------------------------------------------------------
            # DLL
            codeError = (np.sqrt(iEarly**2 + qEarly**2) - np.sqrt(iLate**2 + qLate**2)) / \
                (np.sqrt(iEarly**2 + qEarly**2) + np.sqrt(iLate**2 + qLate**2))
            
            # Update NCO code
            codeNCO += self.dllTau2 / self.dllTau1 * (codeError - oldCodeError) \
                + codeError * (self.pdiCode / self.dllTau1)
            
            oldCodeError = codeError

            codeFrequency = self.signal.code_freq - codeNCO

            # -----------------------------------------------------------------
            # PLL
            carrierError = np.arctan(qPrompt / iPrompt) / 2.0 / np.pi

            # Update NCO carrier
            carrierNCO += self.pllTau2 / self.pllTau1 * (carrierError - oldCarrierError) \
                + carrierError * (self.pdiCarrier / self.pllTau1)

            oldCarrierError = carrierError

            carrierFrequency = self.init_freq + carrierNCO

            # -----------------------------------------------------------------
            # FLL
            # TODO

            # Save variables
            self.absoluteSample.append(signal_file.getCurrentSampleIndex())
            self.codeFrequency.append(codeFrequency)
            self.carrierFrequency.append(carrierFrequency)
            self.codeError.append(codeError)
            self.codeNCO.append(codeNCO)
            self.carrierError.append(carrierError)
            self.carrierNCO .append(carrierNCO)
            self.iEarly.append(iEarly)
            self.qEarly.append(qEarly)
            self.iPrompt.append(iPrompt)
            self.qPrompt.append(qPrompt)
            self.iLate.append(iLate)
            self.qLate.append(qLate)
            self.codePhase.append(remCodePhase)

        # TODO Should pre-allocate all the arrays and put this in numpy arrays right away
        self.iPrompt = np.array(self.iPrompt)

        signal_file.closeFile()

        return


class TrackingAbstract(ABC):

    @abstractmethod
    def __init__(self, rfConfig:RFSignal):
        self.rfConfig  = rfConfig

        self.remCodePhase  = 0.0
        self.codeFrequency = np.nan
        self.codePhaseStep = np.nan
        self.codeNCO       = np.nan

        self.carrierFrequency = np.nan
        self.codeFrequency    = np.nan

        self.frequencyError = np.nan
        self.frequencyNCO   = np.nan
        
        self.code = []
        self.chunck = -1
        self.iSignal = []
        self.qSignal = []
    
    # -------------------------------------------------------------------------
    # ACSTRACT METHODS
    
    @abstractmethod
    def run(self):
        pass

    @abstractmethod
    def delayLockLoop(self):
        pass

    @abstractmethod
    def phaseLockLoop(self):
        pass

    # -------------------------------------------------------------------------
    # METHODS

    def getCorrelator(self, correlatorSpacing):

        idx = np.ceil(np.linspace(self.remCodePhase + correlatorSpacing, \
                self.chunck * self.codePhaseStep + self.remCodePhase + correlatorSpacing, \
                self.chunck, endpoint=False)).astype(int)
        tmpCode = self.code[idx]

        iCorr  = np.sum(tmpCode  * self.iSignal)
        qCorr  = np.sum(tmpCode  * self.qSignal)

        return iCorr, qCorr

    # -------------------------------------------------------------------------
    
    def setInitialValues(self, estimatedFrequency, estimatedCode):
        self.carrierFrequency = estimatedFrequency
        self.codeFrequency    = estimatedCode

    # -------------------------------------------------------------------------

    def setSatellite(self, svid):
        self.svid = svid
        self.code = self.signalConfig.getCode(svid, self.samplingFrequency)
        
        return 

class Tracking_EPL(TrackingAbstract):

    def __init__(self, signalConfig:GNSSSignal, rfConfig:RFSignal):
        super().__init__(signalConfig, rfConfig)
        self.signalConfig = signalConfig
        
        # Load parameters
        config = configparser.ConfigParser()
        config.read(signalConfig.configFile)
        
        self.pdiCode    = config.getfloat('TRACKING', 'pdi_code')
        self.pdiCarrier = config.getfloat('TRACKING', 'pdi_carrier')

        self.correlatorSpacing = config.getfloat('TRACKING', 'correlator_spacing')

        self.dllDumpingRatio      = config.getfloat('TRACKING', 'dll_dumping_ratio')
        self.dllNoiseBandwidth    = config.getfloat('TRACKING', 'dll_noise_bandwidth')
        self.dllLoopGain          = config.getfloat('TRACKING', 'dll_loop_gain')

        self.pllDumpingRatio      = config.getfloat('TRACKING', 'pll_dumping_ratio')
        self.pllNoiseBandwidth    = config.getfloat('TRACKING', 'pll_noise_bandwidth')
        self.pllLoopGain          = config.getfloat('TRACKING', 'pll_loop_gain')

        self.dllTau1, self.dllTau2 = self.getLoopCoefficients(self.dllNoiseBandwidth, \
            self.dllDumpingRatio, self.dllLoopGain)
        self.pllTau1, self.pllTau2 = self.getLoopCoefficients(self.pllNoiseBandwidth, \
            self.pllDumpingRatio, self.pllLoopGain)

        self.correlatorSpacing = [-self.correlatorSpacing, 0.0, self.correlatorSpacing]

        return

    def run(self, rfData):

        # Updates parameters
        self.codePhaseStep = self.signalConfig.codeFrequency - self.codeNCO / self.rfConfig.samplingFrequency
        chunck = int(np.ceil((self.signal.code_bit - self.remCodePhase) / self.codePhaseStep))

        # Generate replica and mix signal
        carrierSignal = np.exp(1j * temp[:chunck]) * rfData
        iSignal = np.real(carrierSignal)
        qSignal = np.imag(carrierSignal)

        # Build correlators (Early-Prompt-Late)
        iEarly , qEarly  = self.getCorrelator(self.correlatorsSpacing[0])
        iPrompt, qPrompt = self.getCorrelator(self.correlatorsSpacing[1])
        iLate  , qLate   = self.getCorrelator(self.correlatorsSpacing[2])
        
        # Delay Lock Loop (DLL)
        self.delayLockLoop(iEarly, qEarly, iLate, qLate)
        
        # Phase Lock Loop (PLL)
        self.phaseLockLoop(iPrompt, qPrompt)

        # Get remaining phase
        idx = np.linspace(self.remCodePhase, chunck * self.codePhaseStep + self.remCodePhase, \
                          chunck, endpoint=False)
        self.remCodePhase = idx[chunck-1] + self.codePhaseStep - self.signal.code_bit

        return

    def generateReplica(self):
        # -----------------------------------------------------------------
        # Generate carrier replica and mix to remove frequency shift
        # We use (chunck+1) and not (chunck) because we want one more to
        # estimate the remaining of the carrier phase
        chunck = int(np.ceil((self.signalConfig.codeBits - self.remCodePhase) / self.codePhaseStep))
        time = np.arange(0, chunck+1) / self.rfConfig.samplingFrequency
        temp = -(self.carrierFrequency * 2.0 * np.pi * time) + remCarrierPhase

        remCarrierPhase = temp[chunck] % (2 * np.pi)
        
        carrierSignal = np.exp(1j * temp[:chunck]) * rawSignal
        iSignal = np.real(carrierSignal)
        qSignal = np.imag(carrierSignal)
    
    def delayLockLoop(self, iEarly, qEarly, iLate, qLate):

        newCodeError = (np.sqrt(iEarly**2 + qEarly**2) - np.sqrt(iLate**2 + qLate**2)) / \
                    (np.sqrt(iEarly**2 + qEarly**2) + np.sqrt(iLate**2 + qLate**2))
            
        # Update NCO code
        self.codeNCO += self.config.dllTau2 / self.config.dllTau1 * (newCodeError - self.codeError)
        self.codeNCO += self.config.pdiCode / self.config.dllTau1 * newCodeError
        
        self.codeError = newCodeError

        return

    def phaseLockLoop(self, iPrompt, qPrompt):

        newFrequencyError = np.arctan(qPrompt / iPrompt) / 2.0 / np.pi

        # Update NCO frequency
        self.frequencyNCO += self.pllTau2 / self.pllTau1 * (newFrequencyError - self.frequencyError)
        self.frequencyNCO += self.pdiCarrier / self.pllTau1 * newFrequencyError

        self.frequencyError = newFrequencyError

        return
    
    def getLoopCoefficients(self, loopNoiseBandwidth, dumpingRatio, loopGain):
        """
        Compute loop coeficients for PLL and DLL loops. From code of [Borre, 2007].

        Args:
            loopNoiseBandwidth (float): Loop Noise Bandwith parameter
            dumpingRatio (float): Dumping Ratio parameter, a.k.a. zeta
            loopGain (float): Loop Gain parameter
        
        Returns
            tau1 (float): Loop filter coeficient (1st)
            tau2 (float): Loop filter coeficient (2nd)

        Raises:
            None
        """

        Wn = loopNoiseBandwidth * 8.0 * dumpingRatio / \
            (4.0 * dumpingRatio**2 +1)
        
        tau1 = loopGain / Wn**2
        tau2 = 2.0 * dumpingRatio / Wn

        return tau1, tau2

File: core/c_functions/tracking.c

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <complex.h>
#include <stdbool.h>
#include <assert.h>

// GPS' definition of Pi
#define PI 3.1415926535898

void readValues(complex double * rfdata, int rfsize){
    // Read values
    for(int i=0; i < 8; i++){
        printf("%f + %fi \n", creal(rfdata[i]), cimag(rfdata[i]));
    }
} 

// --------------------------------------------------------------------------------------------------------------------
// TODO Coud be simplified by just recomputing the value of time in for loop instead of giving an array.
/*
@brief
@param time              Array of length `size+1` with time values ranging from 0 to (size+1) / sampling frequency 
@param size              Length of the arrays 
@param carrierFrequency  Carrier frequency of the signal (i.e. remaining due to Doppler effect)
@param remCarrierPhase   Remaining carrier phase from last loop (initial value is 0.0)
@param r_remCarrierPhase Return of the remCarrierPhase
@param r_replica         Return of the replica array of length `size`
@return void.
*/
void generateReplica(double* time,
                     size_t size,
                     double carrierFrequency,
                     double remCarrierPhase,
                     double* r_remCarrierPhase,
                     complex double* r_replica) 
{
    double temp;

    for (size_t i=0; i < size; i++)
    {
        temp = -(carrierFrequency * 2.0 * PI * time[i]) + remCarrierPhase;
        r_replica[i] = cexp(I * temp);
        //printf("%f\n", temp);
        //printf("%f + i%f\n", creal(r_replica[i]), cimag(r_replica[i]));
    }
    temp = -(carrierFrequency * 2.0 * PI * time[size]) + remCarrierPhase; //0.0094247779607694003

    *r_remCarrierPhase = fmod(temp, 2*PI);

    return;
}

// --------------------------------------------------------------------------------------------------------------------

/*
@brief Perform a correlation operation between the code and I/Q signal.
@param iSignal           Array with the real part of the signal.
@param qSignal           Array with the imaginary part of the signal.
@param code              Array with the PRN code.
@param size              Size of the arrays (iSignal, qSignal, code, iCorr, qCorr).
@param codeStep          The step size between the code frequency and the sampling frequency.
@param remCodePhase      Remaining code phase from last loop.
@param correlatorSpacing Correlator spacing in code chip.
@param r_iCorr             Array with correlation result for the real part.
@param r_qCorr             Array with correlation result for the imaginary part.
@return void.
*/
void getCorrelator(double* iSignal,
                   double* qSignal,
                   int* code,
                   size_t size, // samplesRequired
                   double codeStep,
                   double remCodePhase, 
                   double correlatorSpacing,
                   double* r_iCorr,
                   double* r_qCorr)
{
    size_t codeIdx = 0;
    
    double start = remCodePhase + correlatorSpacing;
    double stop = size * codeStep + remCodePhase + correlatorSpacing;
    double step = (stop - start) / size;

    *r_iCorr = 0.0;
    *r_qCorr = 0.0;
    for (size_t idx=0; idx < size; idx++)
    {
        codeIdx = ceil(start + step * idx);

        *r_iCorr += code[codeIdx] * iSignal[idx];
        *r_qCorr += code[codeIdx] * qSignal[idx];
    }

    return;
}

// --------------------------------------------------------------------------------------------------------------------

/*
@brief
@param
@return void.
*/
void generateCarrier(complex double* rfData,
                     complex double* replica,
                     size_t size,
                     double* r_iSignal,
                     double* r_qSignal)
{
    complex double temp;

    for (size_t i=0; i < size; i++)
    {
        temp = rfData[i] * replica[i];

        r_iSignal[i] = creal(temp);
        r_qSignal[i] = cimag(temp);
    }

    return;
}

// --------------------------------------------------------------------------------------------------------------------

/*
@brief
@param
@return void.
*/
void delayLockLoop(double iEarly,
                   double qEarly,
                   double iLate,
                   double qLate,
                   double dllTau1,
                   double dllTau2,
                   double pdiCode,
                   double codeNCO,
                   double codeError,
                   double codeFrequency,
                   double* r_codeNCO,
                   double* r_codeError,
                   double* r_codeFrequency)
{
    double newCodeError = (sqrt(pow(iEarly, 2) + pow(qEarly, 2)) -
                           sqrt(pow(iLate,  2) + pow(qLate,  2))) /
                          (sqrt(pow(iEarly, 2) + pow(qEarly, 2)) +
                           sqrt(pow(iLate,  2) + pow(qLate,  2)));

    *r_codeNCO  = codeNCO;
    *r_codeNCO += dllTau2 / dllTau1 * (newCodeError - codeError);
    *r_codeNCO += pdiCode / dllTau1 *  newCodeError;

    *r_codeError = newCodeError;

    *r_codeFrequency = codeFrequency - *r_codeNCO;

    return;
}

// --------------------------------------------------------------------------------------------------------------------

/*
@brief
@param
@return void.
*/
void phaseLockLoop(double iPrompt,
                   double qPrompt,
                   double pllTau1,
                   double pllTau2,
                   double pdiCarrier,
                   double carrierNCO,
                   double carrierError,
                   double initialFrequency,
                   double* r_carrierNCO,
                   double* r_carrierError,
                   double* r_carrierFrequency)
{
    double newCarrierError = atan(qPrompt / iPrompt) / 2.0 / PI;

    *r_carrierNCO  = carrierNCO;
    *r_carrierNCO += pllTau2 / pllTau1 * (newCarrierError - carrierError);
    *r_carrierNCO += pdiCarrier / pllTau1 * newCarrierError;

    *r_carrierError = carrierError;

    *r_carrierFrequency = initialFrequency + *r_carrierNCO;

    return;
}

// --------------------------------------------------------------------------------------------------------------------

/*
@brief
@param
@return void.
*/
void getLoopCoefficients(double loopNoiseBandwidth,
                         double dumpingRatio,
                         double loopGain,
                         double* r_tau1,
                         double* r_tau2)
{
    double wn = loopNoiseBandwidth * 8.0 * dumpingRatio / (4.0 * pow(dumpingRatio, 2) + 1);

    *r_tau1 = loopGain / pow(wn, 2);
    *r_tau2 = 2.0 * dumpingRatio / wn;

    return;
}

// --------------------------------------------------------------------------------------------------------------------

bool test_generateReplica(){
    /*
    Unit test function of generateReplica.
    */
    int size = 5;
    int samplingFrequency = 1e7;
    double carrierFrequency = -1500.0;
    double remCarrierPhase = 0.0;
    double time[size+1];
    double r_remCarrierPhase = 0.0;
    complex double r_replica[size];
    complex double r_replica_truth[size];
    float epsilon = 1e-8;
    bool success = true;
    
    // init 
    for(int i=0; i < size+1; i++){
        time[i] = (double) i / samplingFrequency;
    }
    for(int i=0; i < size; i++){
        r_replica[i] = 0.0;
    }

    // test function
    generateReplica(time, size, carrierFrequency, remCarrierPhase, &r_remCarrierPhase, r_replica);

    // Assert
    r_replica_truth[0] = 1 + I * 0.0;
    r_replica_truth[1] = 0.9999995558678348 + I * 0.000942477656548699;
    r_replica_truth[2] = 0.9999982234717338 + I * 0.0018849544759281136;
    r_replica_truth[3] = 0.9999960028128805 + I * 0.002827429620969703;
    r_replica_truth[4] = 0.9999928938932473 + I * 0.0037699022545064132;

    for(int i=0; i < size; i++){
        if(fabs(r_replica_truth[i] - r_replica[i]) > epsilon){
            success = false;
            break;
        }
    }

    assert(success == true);

    return success;
}

/// ===================================================================================================================

int main(){
    printf("Hello World!\n");

    // generateReplica unit test 
    // test_generateReplica();
    
    // getCorrelator unit test
    int size = 5;
    double iSignal[5] = {-5.061244639011612, 9.39905932245856, 5.521358200447387, -3.904872069362163, 17.396453218094308};
    double qSignal[5] = {41.46545312310088, 22.992557140364028, -13.982653668826686, -7.533390612593724, -18.637688038773916};
    int code[5] = {-1, 1, 1, 1, -1};
    double codeStep = 0.1023;
    double remCodePhase = 0.0;
    double correlatorSpacing = 0.5;
    double r_iCorr = 0.0;
    double r_qCorr = 0.0;

    getCorrelator(iSignal, qSignal, code, size, codeStep, remCodePhase, correlatorSpacing, &r_iCorr, &r_qCorr);

}